### ISSUE TITLE
High CPU Usage

### ISSUE THREAD

**Issue by theteachr** [leostera/minttea#38]
I see that running any simple application with `minttea` churns a lot of CPU. I was not able to run the examples from the repo itself, but was able to create a simple project with `minttea`, copy the examples, and run.

![Screenshot 2024-03-03 at 2 28 03 AM](https://github.com/leostera/minttea/assets/64728420/30014b51-a2c4-4802-bb17-d86f1c6dc37d)

The screenshot is from macOS on an M1 chip. But I've also seen high CPU usage on WSL (Ubuntu 22.04) and Linux (Fedora 34).

**Comment by bigspeedfpv** [comment]
Same here - interestingly, nearly the same CPU usage: 
<img width="299" alt="image" src="https://github.com/leostera/minttea/assets/21012310/7da86519-2f0f-4b7b-9a32-021246acf2ed">

**Comment by leostera** [comment]
Hi folks! This looks like an issue on Riot. I'll transfer the issue over there.

**Comment by chshersh** [comment]
I actually think that solving this issue may fix

* #34 

I believe that due to high CPU usage, the CPU starts to throttle and thus we observe this huge slowdown.

**Comment by leostera** [comment]
@theteachr and @chshersh could you check running minttea against the latest riot from repo? We solved some busy-waiting issues that should improve the resource usage quite a bit.

**Comment by theteachr** [comment]
@leostera I see a significant decrease in the CPU usage: From 250% to around 8%. These numbers are from a base model Mac Studio (10 Core CPU, Apple M1).

**Comment by leostera** [comment]
Sweet! That ~8% is what I expect from the timer resolution we currently have (microseconds). Lower resolution (eg. milliseconds) could help us drop that to ~3%. We may make this configurable at a later stage for different workloads.

If you feel comfortable with this CPU usage for now, feel free to close the issue and we can reopen if anything goes haywire 🙏🏼

**Comment by chshersh** [comment]
@leostera This is awesome sauce! I think the situation has improved for me but not entirely.

I still see an issue from #34 but it happens not so frequently to me. Previously, I had to restart my app 7 times to avoid hanging. Now I need to restart it 7 times to observe hanging, which is an improvement!

At this point, I started thinking that my `view` function may generate lots of garbage, and the GC struggles to process everything at 60 FPS (default).

I wonder if Mint Tea can add something like `has_diff : 'model -> 'model -> bool` to `t`

https://github.com/leostera/minttea/blob/3e386136d315c60c16ea39b94789fbb7761003f5/minttea/app.ml#L1-L5

So that it can avoid calling `view` entirely if the `model` didn't change unlike the current implementation where we compare already rendered buffer:

https://github.com/leostera/minttea/blob/3e386136d315c60c16ea39b94789fbb7761003f5/minttea/renderer.ml#L26

<hr/>

Alternatively, exposing the `fps` setting from `start` may help app developers reduce their resource usage (I don't really need 60 fps)

https://github.com/leostera/minttea/blob/3e386136d315c60c16ea39b94789fbb7761003f5/minttea/minttea.ml#L9-L26
